src/renderers/WebGLRenderer.js

```diff
@@ -49,6 +49,10 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 	this.sortObjects = true;
 
+	// user-defined clipping
+
+	this.userClippingPlanes = [];
```

Comment:
> As far as I understand, this is mainly uniforms. Wouldn't it be better to set this array in the material instead?

This array holds `THREE.Plane` objects in world space. 

Right now, clipping planes completely mask all rendering - not for a single object, not just for meshes: Points, sprites, meshes, whatever with a built-in shader. Clipping for shadow occlusion must be optional, as discussed above. 

The state is global and the resulting uniform value is shared between all materials - similar as we do for the lights (please note that the implementation of the update mechanics are still half-baked in the code here right now - I'm still hardening it).

There are several use cases where it would be cool to allow it on a per object basis. I don't think it really belongs onto the material, but it would be the "reasonably close approximation right now" to avoid issues like #8379 and keep this PR focussed (I guess that's why you are suggesting it?). 

Again, as with lighting, there is a conversion to view space in the renderer. This way, there are no problems when running on a platform with reduced fragment shader precision (floats scale nicely around the origin. When they are offset, the exponent gets defined by the translation and there are fewer bits to work with for fine-grained stuff. The nice thing about view space is, that we carry the origin around with us and everything closely visible is near the origin and gets decent precision). So there's a dependence on the camera and putting clipping planes on the material may end up a complication rather than a simplification.

Clipping planes effectively restrict the view frustum, so it's worthwhile to consider putting them on the camera object.

The dream API would have global clipping planes on the camera to mask the viewport  - ones that do not affect occluders when rendering a shadow map (just because you can't see the object doesn't mean you can't see its shadow). Then there could be additional ones on an object that semantically cut the object and therefore quite logically also its shadow.

One could then also unify frustum culling (should be a separate step) by always having six entries minimum - the clipping planes of the camera's view frustum.